scripts/revng-check-conventions

```diff
@@ -100,134 +109,244 @@ while [[ $# > 0 ]]; do
-        )"
-    $GREP "^\s*#include <.*\.h>" $CXX_FILES | cat
+    readarray -t CXX_FILES < <(printf '%s\n' "${FILES[@]}" | \
+        grep -v \
+            -e '^runtime/support\.h$' \
+            -e '^include/revng/Runtime/.*$' \
+            -e '^include/revng/PipelineC/.*$' \
+            -e '^include/revng/Support/Assert\.h$' \
+            -e '^include/revng/Support/ClassSentinel\.h$' \
+            -e '\.c$' \
+    )
+    if [[ ${#CXX_FILES[@]} -ne 0 ]]; then
+        cat <($GREP "^\s*#include <.*\.hpp>" "${FILES[@]}") \
+            <($GREP "^\s*#include <.*\.h>" "${CXX_FILES[@]}") \
+            | cat_header "Includes should never user <..> except for C++ standard includes:"
+    else
+        $GREP "^\s*#include <.*\.hpp>" "${FILES[@]}" | \
+            cat_header "Includes should never user <..> except for C++ standard includes:"
+    fi
 
     # Parenthesis at the end of line (except for raw strings)
-    $GREP "(\$" $FILES | grep -v 'R"LLVM.*(' | cat
+    $GREP "(\$" "${FILES[@]}" | grep -v 'R"LLVM.*(' | cat_header "Parenthesis at the end of line:"
 
     # Ban whitespaces at the end of a line
-    $GREP " \$" | cat
+    $GREP " \$" "${FILES[@]}" | cat_header "Whitespace at the end of line:"
 
     # Ban tabs altogether
-    $GREP -P "\t" | cat
+    $GREP -P "\t" "${FILES[@]}" | cat_header "Tabs present:"
 
     # Ensure all files end with a newline
-    git ls-files | while read FILE; do
-        if test -s "$FILE" && ! test $(tail -c 1 "$FILE" | base64) == "Cg=="; then
+    while read -r FILE; do
+        if [[ -s "$FILE" && ! $(tail -c 1 "$FILE" | base64) == "Cg==" ]]; then
             echo "$FILE does not end with a newline"
         fi
-    done
+    done < <($GIT_LS_FILES) | cat_header "Files that don't end in a newline:"
 
     # Things should never be at the beginning of a line
-    for REGEXP in '\*>' '/[^/\*]' ':[^:\(]*)' '==' '\!=' '<[^<]' '>' '>=' '<=' '//\s*WIP' '#if\s*[01]'; do
-        $GREP "^\s*$REGEXP" $FILES | cat
-    done
+    REGEXPS=('\*>' '/[^/\*]' ':[^:\(]*)' '==' '\!=' '<[^<]' '>' '>=' '<=' '//\s*WIP' '#if\s*[01]')
+    for REGEXP in "${REGEXPS[@]}"; do
+        $GREP "^\s*$REGEXP" "${FILES[@]}"
+    done | cat_header "Things that should never be at the beginning of a line:"
 
     # Check there are no static functions in C++ header files
-    for FILE in $CXX_FILES; do
+    for FILE in "${CXX_FILES[@]}"; do
         if [[ $FILE == *h ]]; then
             $GREP -H '^static\b[^=]*$' "$FILE" | cat
-            head -n1 "$FILE" |  grep -E '^#pragma once$' > /dev/null || echo "$FILE: header does not start with #pragma once"
+            head -n1 "$FILE" |  grep -E '^#pragma once$' > /dev/null || \
+                echo "$FILE: header does not start with #pragma once"
         fi
     done
 
 }
 
-if test $( ( run_clang_format && run_black && run_cmake_format && run_revng_checks | sort -u ) |& tee /dev/stderr | wc -c ) -ne 0; then
+tmp_file=$(mktemp)
+exec 1> >(tee "$tmp_file" > /dev/stdout)
```

Comment:
Comment what's going on here, very obscure.

Reply:
Added a comment to clarify